qualtran/qref_interop/_bloq_to_qref.py

```diff
@@ -0,0 +1,311 @@
+
+    There are several Bloq classes, however, they all share common set of atributes.
+    This function is used to extract them, so that we don't have to duplicate logic
+    for each Bloq subtype.
+
+    Args:
+        bloq: Bloq for which information should be extracted.
+        name: Optional override for name. Setting it to None will cause some reasonable
+            name (depending on bloq type) to be inferred. The typical use of non-trivial
+            name is when a better name is known from the parent Bloq.
+
+    Returns:
+        A dictionary that can be unpacked into arguments of RoutineV1 initializer.
+    """
+    ports = [port for reg in bloq.signature for port in _ports_from_register(reg)]
+    local_variables = {}
+    for port in ports:
+        if not _is_symbol_or_int(str(port.size)):
+            local_variable_name = f"{port.name}_size"
+            local_variables[local_variable_name] = port.size
+            port.size = local_variable_name
+
+    if name is None:
+        name = bloq.__class__.__name__
+        try:
+            if bloq.uncompute:
+                name += "_uncompute"
+        except AttributeError:
+            pass
+        try:
+            if bloq.is_adjoint:
+                name += "_adjoint"
+        except AttributeError:
+            pass
+    input_params = _extract_input_params(bloq, ports, local_variables)
+
+    attributes = {
+        "name": name,
+        "type": _bloq_type(bloq),
+        "ports": ports,
+        "resources": _import_resources(bloq),
+        "input_params": input_params,
+    }
+    if len(local_variables) > 0:
+        attributes["local_variables"] = local_variables
+    return attributes
+
+
+def _bloq_instance_name(instance: BloqInstance) -> str:
+    """Infer unique (but readable) name for a BloqInstance.
+
+    Child Bloqs in CompositeBloq (and some other places) are stored as BloqInstances,
+    which combine a Bloq with a unique ID. When converting such BloqInstance to Bartiq
+    RoutineV1, the ID has to be incorporated into the name, because otherwise one could
+    get several siblings having the same name.
+    """
+    return f"{_bloq_type(instance.bloq)}_{instance.i}"
+
+
+def bloq_to_qref(obj) -> SchemaV1:
+    """Converts Bloq to QREF SchemaV1 object."""
+    return SchemaV1(version="v1", program=bloq_to_routine(obj))
+
+
+@singledispatch
+def bloq_to_routine(obj, name: Optional[str] = None):
+    """Import object from Qualtran by converting it into corresponding QREF RoutineV1 object.
+
+    Args:
+        obj: object to be imported. Can be either Bloq or BloqInstance.
+        name: optional name override. This can be useful e.g. if you are converting
+            CompositeBloq (which do not store meaningful names in Qualtran) and
+            know some good name for it.
+
+    Return:
+        A Bartiq object corresponding to the source Qualtran object. For both Bloqs
+        and BloqInstances the returned object is of type RoutineV1.
+
+    """
```

Comment:
If I understand correctly, as part of the conversion; we are not doing any bloq decompositions and the expectation is that the user should flatten out a bloq that has a decomposition into a `CompositeBloq` at an appropriate level and then convert the `CompositeBloq` to qref's `SchemaV1`; the conversion would not have any further information about the structure of the subbloqs (eg: info about decomposing a bloq that appears in the composite bloq but is not decomposed). Is this understanding correct?

Reply:
sounds good, I think this is fine as long as we're clear that a `RoutineV1` is analogous to one `CompositeBloq`